pkg/operator/controller/gatewayapi-upgradeable/controller.go

```diff
@@ -0,0 +1,183 @@
+		},
+		DeleteFunc: func(deleteEvent event.DeleteEvent) bool { return false }, // Ignore delete events
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// Watch for updates to the "admin-gates" ConfigMap to check if the admin key changes
+			configmapOld := e.ObjectOld.(*corev1.ConfigMap)
+			configmapNew := e.ObjectNew.(*corev1.ConfigMap)
+			if configmapOld.Name != operatorcontroller.AdminGatesConfigMapName().Name || configmapOld.Namespace != operatorcontroller.AdminGatesConfigMapName().Namespace {
+				return false
+			}
+			oldVal, oldExists := configmapOld.Data[GatewayAPIAdminKey]
+			newVal, newExists := configmapNew.Data[GatewayAPIAdminKey]
+			return oldExists != newExists || oldVal != newVal
+		},
+		GenericFunc: func(genericEvent event.GenericEvent) bool { return false }, // Ignore generic events
+	}
+
+	toAdminGateConfigMap := func(_ context.Context, _ client.Object) []reconcile.Request {
+		return []reconcile.Request{{
+			NamespacedName: operatorcontroller.AdminGatesConfigMapName(),
+		}}
+	}
+	// Define the CRD predicate
+	crdPredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.(*apiextensionsv1.CustomResourceDefinition).Spec.Group == gatewayapiv1.GroupName
+	})
+
+	//TODO: too many args to c.watch and not enough to source.Kind
+	if err := c.Watch(source.Kind(operatorCache, &corev1.ConfigMap{}),
+		handler.EnqueueRequestsFromMapFunc(toAdminGateConfigMap),
+		adminGateFilter); err != nil {
+		return nil, err
+	}
+
+	// Watch for CRD updates using the reconciler cache and apply the CRD predicate
+	if err := c.Watch(source.Kind(reconciler.cache, &apiextensionsv1.CustomResourceDefinition{}),
+		&handler.EnqueueRequestForObject{},
+		crdPredicate); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// reconciler struct holds the client and cache attributes
+type reconciler struct {
+	client client.Client
+	cache  cache.Cache
+}
+
+// Config struct holds the cache information
+type Config struct {
+	//namespace referes to namespace where admin gates controller resides
+	Namespace string
+	//cache should be a namespace global cache
+	Cache cache.Cache
+}
+
+// Reconcile function implements the logic to check conditions and manage the admin gate
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+	adminGateConditionExists, err := r.adminGateConditionExists(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to determine if UnservableInFutureVersionsRoute routes exist: %w", err)
+	}
+	if adminGateConditionExists {
+		if err := r.addAdminGate(ctx); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to add admin gate: %w", err)
+		}
+	} else {
+		if err := r.removeAdminGate(ctx); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to remove admin gate: %w", err)
+		}
+	}
```

Comment:
I suppose that the decision about adding or removing a data field to/from the admin in gate should be based on the presence of GWAPI CRDs which I don't see here.